### PR TITLE
Add container_test bazel rule

### PR DESCRIPTION
**Pull request by nkubala**
This rule wraps the [container structure test](https://github.com/GoogleCloudPlatform/container-structure-test) framework.

cc @dlorenc @mattmoor

### CODE REVIEW

File: container/test.bzl [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+def _impl(ctx):
+    flat_image = ctx.attr.image
+
+    config_runfiles = []
+    config_str = ""
+
+    for config_file in ctx.files.configs:
+        config_runfiles.append(config_file)
+        config_str = config_str + '$(pwd)/' + config_file.short_path + ' '
```
Comment (dlorenc) [inline container/test.bzl]: A list comprehension with string formatting and ' '.join might be easier to read here.

File: container/test.bzl [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    image_name = "bazel/%s:%s" % (ctx.attr.image.label.package, ctx.attr.image.label.name)
```
Comment (dlorenc) [inline container/test.bzl]: I'm not sure if there's a better mechanism for getting the loaded images name here, but having one would be nice. Maybe the image loader could output something?

Comment (mattmoor) [reply container/test.bzl]: Users can override the `bazel/` portion. :(

File: testdata/BUILD [hunk 3]
```diff
@@ -275,6 +276,12 @@ container_flatten(
     image = ":link_with_files_base",
 )
 
+container_test(
+    name = "structure test",
+    image = ":link_with_files_base",
+    configs = ["foo.yaml"],
```
Comment (mattmoor) [inline testdata/BUILD]: Did you miss `foo.yaml`?

File: container/test.bzl [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        "image": attr.label(
+            executable = True,
+            mandatory = True,
+            cfg = "target",
+        ),
+        "configs": attr.label_list(
+            mandatory = True,
+            allow_files = True,
+        ),
+        "verbose": attr.bool(
```
Comment (mattmoor) [inline container/test.bzl]: Can this be sensitive to `--test_output=errors`?  @damienmg

File: container/test.bzl [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            default = False,
+            mandatory = False,
+        ),
+        "driver": attr.string(
```
Comment (mattmoor) [inline container/test.bzl]: You aren't plumbing this into incremental load.  My recommendation would be: don't rely on incremental load.

File: container/structure-test.sh.tpl [hunk 6]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+set -ex
+
+./%{load_statement}
```
Comment (mattmoor) [inline container/structure-test.sh.tpl]: This is a glorious PITA to get right.  You should look at the `guess_runfiles()` logic sprinkled ~everywhere. 
 However, I'd recommend against invoking the incremental loader.

tl;dr This should `docker load -i <tarball>` and depend on the `.tar` for now (see `#3` below)

... but I'd still like to see:
1. ~`container_fs_test`: which under the hood uses `container_flatten` and checks for files in the tarball (doesn't need Docker).
1. ~`container_cfg_test`: which does the same, but consumes the `config.json` emitted (again, no Docker).
1. ~`container_exec_test`: which loads the `image` into the configured `ctx.attr.driver` and executes the configured command.

File: container/container.bzl [hunk 7]
```diff
@@ -145,3 +149,10 @@ py_library(
       remote = "https://github.com/google/subpar",
       commit = "7e12cc130eb8f09c8cb02c3585a91a4043753c56",
     )
+
+  if "structure_test" not in excludes:
+    native.git_repository(
+      name = "container_structure_test",
+      remote = "https://GoogleCloudPlatform/container-structure-test.git",
+      tag = "STRUCTURE_TEST_RELEASE",
```
Comment (mattmoor) [inline container/container.bzl]: drop quotes

File: container/structure-test.sh.tpl (hunk 6 above)
Comment (mattmoor) [reply container/structure-test.sh.tpl]: Frankly...

`#1` could be sugar around `container_flatten` and a `tar_test` rule we upstream to `pkg_tar`.
`#2` could be sugar around `container_flatten` and a `json_test` rule we upstream somewhere.

File: container/test.bzl [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A wrapper around container structure tests for Bazel.
+
+This rule feeds a built image and a set of config files
+to the container structure test framework."
+"""
+
+load("//container:flatten.bzl", "container_flatten")
```
Comment (mattmoor) [inline container/test.bzl]: You got me excited, but this isn't used :-/

File: container/test.bzl (hunk 2 above)
Comment (mattmoor) [reply container/test.bzl]: A trick: feed the image into `container_bundle` and name it w/e you want

File: container/test.bzl (hunk 8 above)
Comment (nkubala) [reply container/test.bzl]: yeah I had grandiose plans to add this as an option, but didn't get that far. we can look into supporting it later on but I'll remove for now

File: container/test.bzl (hunk 2 above)
Comment (nkubala) [reply container/test.bzl]: good idea, I'll give that a shot

File: container/structure-test.sh.tpl (hunk 6 above)
Comment (nkubala) [reply container/structure-test.sh.tpl]: what's the reasoning behind not wanting to do the incremental load here? I'm fine with changing it to do an explicit docker load, just trying to understand.

I agree with having a few macros to support different use cases, but I just wasn't sure which ones we really cared about at this point in time, i.e. which will actually be used. `container_flatten`ing everything before feeding into the test seems like a generally good practice, but if we're not actually using flat images with this right now it might be overkill. maybe this should happen in a separate PR? WDYT?

File: container/test.bzl (hunk 5 above)
Comment (nkubala) [reply container/test.bzl]: this is actually referring to the driver that the test framework itself will use to run the tests: currently we support using docker (the default driver), and then a "tar" driver for manipulating file systems without the need for a working docker daemon (currently doesn't support command tests). the load needs to happen regardless before this driver does anything, it's just passed as a string flag to the test command.

File: WORKSPACE [hunk 9]
```diff
@@ -135,3 +135,9 @@ git_repository(
 load("@io_bazel_rules_d//d:d.bzl", "d_repositories")
 
 d_repositories()
+
+git_repository(
+    name = "io_bazel_structure_test",
```
Comment (mattmoor) [inline WORKSPACE]: `io_bazel` seems like the wrong namespace.  Why isn't this in the `repositories()` function?

File: container/structure-test.sh.tpl (hunk 6 above)
Comment (mattmoor) [reply container/structure-test.sh.tpl]: I generally regard the incremental load as (slightly) more brittle, it isn't always faster (e.g. if there are lots of images in the daemon), and I doubt we'll ever try to make it "pluggable" in the same sense as `driver`.

File: WORKSPACE (hunk 9 above)
Comment (nkubala) [reply WORKSPACE]: I was just stupidly pattern matching, but you're right it doesn't really belog here. moved

File: container/test.bzl (hunk 1 above)
Comment (nkubala) [reply container/test.bzl]: good idea

File: container/test.bzl [hunk 10]
```diff
@@ -82,3 +85,22 @@ container_test = rule(
     test = True,
     implementation = _impl,
 )
+
+
+def container_test(name, image, configs, driver=None, verbose=None):
+    intermediate_image_name = "%s.intermediate" % image
+
+    container_bundle(
```
Comment (mattmoor) [inline container/test.bzl]: `# Give the image a predictable name when loaded`

File: container/test.bzl [hunk 11]
```diff
@@ -32,7 +34,8 @@ def _impl(ctx):
           "%{configs}": config_str,
           "%{workspace_name}": ctx.workspace_name,
           "%{test_executable}": ctx.executable._structure_test.short_path,
-          "%{image}": image_name,
+          # "%{image}": image_name,
```
Comment (mattmoor) [inline container/test.bzl]: Remove commented code

File: container/test.bzl [hunk 12]
```diff
@@ -82,3 +85,22 @@ container_test = rule(
     test = True,
     implementation = _impl,
 )
+
+
+def container_test(name, image, configs, driver=None, verbose=None):
+    intermediate_image_name = "%s.intermediate" % image
+
+    container_bundle(
+        name = "intermediate_bundle",
```
Comment (mattmoor) [inline container/test.bzl]: I don't think you can pick such a stable name, unfortunately.

File: container/test.bzl [hunk 13]
```diff
@@ -82,3 +85,22 @@ container_test = rule(
...
+        images = {
+            intermediate_image_name: image,
+        }
+    )
+
+    _container_test(
+        name = name,
+        image = intermediate_image_name,
```
Comment (mattmoor) [inline container/test.bzl]: This rule doesn't depend on the `container_bundle`.

File: container/test.bzl (hunk 11 above)
Comment (nkubala) [reply container/test.bzl]: sorry, I pushed a WIP commit to this branch because I switched machines, should have let you know. fixed now

File: container/test.bzl (hunk 4 above)
Comment (nkubala) [reply container/test.bzl]: how would I address this? not really sure where that flag should get handled

**Comment by mattmoor** [comment]
Bazel seems upset: https://ci.bazel.io/blue/organizations/jenkins/PR%2Frules_docker/detail/rules_docker/226/tests

**Comment by damienmg** [comment]
We had a discussion offline with @mattmoor and he told me you had a test that needs to run against the docker daemon.

We have slaves on ci.bazel.io that can talk to the docker daemon. Simply use `docker` as `node` in https://github.com/bazelbuild/rules_docker/blob/master/.ci/rules_docker.json. For instance see https://github.com/bazelbuild/bazel/blob/master/scripts/ci/bazel-slow-tests.json#L3 on how we run the same kind of test in Bazel.

File: contrib/structure-test.sh.tpl [hunk 14]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+set -ex
```
Comment (mattmoor) [inline contrib/structure-test.sh.tpl]: I generally try to avoid `-x` in what's checked in.

File: contrib/test.bzl [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+          "%{image}": image_name,
+          "%{driver}": ctx.attr.driver,
+        },
+        is_executable=True
+    )
+
+    return struct(runfiles=ctx.runfiles(files = [
+            ctx.executable._structure_test,
+            ctx.executable.image_tar,
+            ctx.file.image_tar] +
```
Comment (mattmoor) [inline contrib/test.bzl]: Can you put the close-bracket on its own line?  I think it will make it clearer that you switch from `+`ing individual files vs. lists

File: testdata/BUILD [hunk 16]
```diff
@@ -276,6 +277,12 @@ container_flatten(
     image = ":link_with_files_base",
 )
 
+# container_test(
+#     name = "structure_test",
+#     configs = ["test.yaml"],
+#     image = ":link_with_files_base",
+# )
```
Comment (mattmoor) [inline testdata/BUILD]: Does the `tar` driver need Docker?

I'd actually like to see us exercise a few behaviors of the test framework here (these kind of serve as samples I often link to as well).

Can you create a `tar` driver that should run anywhere, and follow @damienmg guidance for running with Docker on Bazel CI?

File: container/container.bzl [hunk 17]
```diff
@@ -145,3 +148,10 @@ py_library(
       remote = "https://github.com/google/subpar",
       commit = "7e12cc130eb8f09c8cb02c3585a91a4043753c56",
     )
+
+  if "structure_test" not in excludes:
```
Comment (mattmoor) [inline container/container.bzl]: Generally we don't define `contrib` deps here.  Can you create a `repositories()` function in `test.bzl` for this?

What transitive deps does this have?  I also worry that we're introducing an implicit dependency on `rules_go`, which cannot easily be encapsulated into one of these function (further motivation to separate it's setup).

File: testdata/test.yaml [hunk 18]
```diff
@@ -0,0 +1,11 @@
+schemaVersion: 2.0.0
+
+fileExistenceTests:
+  - name: foo
+    path: "/foo"
+    shouldExist: true
+
+fileContentTests:
```
Comment (mattmoor) [inline testdata/test.yaml]: I think I'd like to see us add `exec` tests for the `lang_image` rules.

File: contrib/structure-test.sh.tpl (hunk 14 above)
Comment (nkubala) [reply contrib/structure-test.sh.tpl]: fair

File: contrib/test.bzl (hunk 15 above)
Comment (nkubala) [reply contrib/test.bzl]: done

File: container/container.bzl (hunk 17 above)
Comment (nkubala) [reply container/container.bzl]: Ah yep, I'll pull it out. All of the deps in the container-structure-test are vendored in, and `rules_do` isn't one of them, so I think we should be good here.

File: testdata/BUILD (hunk 16 above)
Comment (nkubala) [reply testdata/BUILD]: Nope, the point of the tar driver is that it explicitly doesn't use Docker (which is why command tests don't currently work with it). I'll add a run of some file tests using the tar driver, and then try and use @damienmg's suggestion to get another run working with Docker.

File: contrib/test.bzl [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+    )
+
+    _container_test(
+        name = name,
+        image_name = intermediate_image_name,
+        image_tar = image_tar_name + ".tar",
+        configs = configs,
+        verbose = verbose,
+        driver = driver,
+        tags = tags,
```
Comment (tejal29) [inline contrib/test.bzl]: Not used

File: contrib/test.bzl [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+        "image_name": attr.string(
+            mandatory = True,
+        ),
+        "configs": attr.label_list(
+            mandatory = True,
+            allow_files = True,
+        ),
+        "verbose": attr.bool(
+            default = False,
+            mandatory = False,
```
Comment (tejal29) [inline contrib/test.bzl]: you can skip mandatory=False.

File: contrib/test.bzl [hunk 21]
```diff
@@ -0,0 +1,144 @@
...
+        ),
+        "driver": attr.string(
+            default = "docker",
+            doc = "Driver to use when running structure tests",
+            mandatory = False,
```
Comment (tejal29) [inline contrib/test.bzl]: I think this needs to be True or set a default. This is used at L51 above.

File: contrib/test.bzl [hunk 22]
```diff
@@ -0,0 +1,144 @@
...
+            allow_files = True,
+            single_file = True,
+        ),
+    },
+    executable = True,
+    test = True,
+    implementation = _impl,
+)
+
+def container_test(name, image, configs, driver=None, verbose=None, tags=[]):
```
Comment (tejal29) [inline contrib/test.bzl]: Can you set default value for driver ? Set one from the allowed list mentioned in the rule.

File: testdata/docker/BUILD [hunk 23]
```diff
@@ -0,0 +1,183 @@
...
+)
+
+container_test(
+    name = "structure_test",
+    configs = ["test.yaml"],
+    image = ":link_with_files_base",
+    tags = ["docker"],
+)
+
+# load("//python:image.bzl", "py_image")
```
Comment (tejal29) [inline testdata/docker/BUILD]: Can you add a todo or description on why this is commented

File: contrib/test.bzl [hunk 24]
```diff
@@ -0,0 +1,144 @@
...
+                ctx.executable.image_tar,
+                ctx.file.image_tar
+            ] +
+            ctx.attr.image_tar.files.to_list() +
+            ctx.attr.image_tar.data_runfiles.files.to_list() +
+            ctx.files.configs,
+        ),
+    )
+
+_container_test = rule(
```
Comment (tejal29) [inline contrib/test.bzl]: Can you add a doc string like this https://github.com/GoogleCloudPlatform/base-images-docker/blob/master/package_managers/download_pkgs.bzl#L72
This will generate documentation using skydoc.

File: contrib/test.bzl [hunk 25]
```diff
@@ -0,0 +1,144 @@
...
+          "%{workspace_name}": ctx.workspace_name,
+          "%{test_executable}": ctx.executable._structure_test.short_path,
+          "%{image}": image_name,
+          "%{driver}": ctx.attr.driver,
+        },
+        is_executable = True
+    )
+
+    return struct(runfiles = ctx.runfiles(files =
+            [
```
Comment (tejal29) [inline contrib/test.bzl]: fix indentation.
```
return struct(runfiles=ctx.runfiles(files=[
                ctx.executable._structure_test,
                ctx.executable.image_tar,
                ctx.file.image_tar
            ] +
            ctx.attr.image_tar.files.to_list() +
            ctx.attr.image_tar.data_runfiles.files.to_list() +
            ctx.files.configs,
        ),
    )
```

I find this online tool handy http://pep8online.com/

File: contrib/test.bzl [hunk 26]
```diff
@@ -0,0 +1,144 @@
...
+        "driver": attr.string(
+            default = "docker",
+            doc = "Driver to use when running structure tests",
+            mandatory = False,
+            values = [
+                "docker",
+                "tar",
+            ],
+        ),
+        "_structure_test": attr.label(
```
Comment (tejal29) [inline contrib/test.bzl]: Is the go binary for structure test?

File: testdata/docker/test.yaml [hunk 27]
```diff
@@ -0,0 +1,11 @@
...
+
+fileExistenceTests:
+  - name: foo
+    path: "/foo"
+    shouldExist: true
+
+fileContentTests:
+  - name: foo
+    path: "/foo"
+    expectedContents: "asdf$"
```
Comment (tejal29) [inline testdata/docker/test.yaml]: Add symlinks existence rule.

File: contrib/test.bzl (hunk 25 above)
Comment (nkubala) [reply contrib/test.bzl]: This repo as a whole needs some formatting love, I'm not sure it adheres to any of the standard python conventions. I'll fix up my file, then later on we might want to add a `pep8` call or something to travis and fix up the rest of the files.

File: contrib/test.bzl (hunk 21 above)
Comment (nkubala) [reply contrib/test.bzl]: it's set to default to `docker` two lines above. this should get selected if the user doesn't pass one in.

File: contrib/test.bzl (hunk 20 above)
Comment (nkubala) [reply contrib/test.bzl]: good call, done

File: contrib/test.bzl (hunk 26 above)
Comment (nkubala) [reply contrib/test.bzl]: yep, this target is the binary built from source

File: contrib/test.bzl (hunk 24 above)
Comment (nkubala) [reply contrib/test.bzl]: done

File: contrib/test.bzl (hunk 21 above)
Comment (tejal29) [reply contrib/test.bzl]: ah then drop mandatory=False.

File: contrib/test.bzl (hunk 26 above)
Comment (tejal29) [reply contrib/test.bzl]: Can we rename this to _structure_test_binary or _structure_test_executable like in the _structure_test.tpl file?

File: testdata/test.yaml (hunk 18 above)
Comment (nkubala) [reply testdata/test.yaml]: yeah I think that's a good use case. I'm going to add tests in a follow up PR, this one is getting a bit cluttered and I'll need to fiddle with the bazel CI a bit which I don't want to do here.

**Comment by nkubala** [comment]
After struggling a bit with the bazel CI, I decided I'm going to address the tests in a follow-up PR, so I don't have to keep tacking on WIP commits onto this one to kick off the CI. I'll talk with @damienmg to get things configured correctly. Should be RFAL

File: testdata/BUILD (hunk 16 above)
Comment (mattmoor) [reply testdata/BUILD]: Can we uncomment these if we make them use the `tar` driver?

**Comment by nkubala** [comment]
@mattmoor please merge when ready

File: testdata/BUILD [hunk 28]
```diff
@@ -20,7 +20,9 @@ load(
     "container_flatten",
     "container_image",
     "container_import",
+    # "container_test",
```
Comment (dlorenc) [inline testdata/BUILD]: remove this commented out line?

**Comment by dlorenc** [comment]
LGTM, needs a rebase then I can merge.
